### ISSUE TITLE
ZeroDivisionError

### ISSUE THREAD

**Issue by Pehnny** [Pehnny/hyperfluorescence#10]
### ZeroDivisionError encountered inside the _hole_electrostatic_energy() method of the Lattice object
```python
Traceback (most recent call last):
  File "/home/tpiron/maxoled/reseau.py", line 429, in _remove_move_hole_events
    index = self._hole_events.index(event)
ValueError: Event(initial=Point(x=15, y=0, z=0), final=Point(x=16, y=0, z=0), tau=0.0, kind=1, particule=2) is not in list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tpiron/maxoled/OLED.py", line 6, in <module>
    oled.operations(recombinations)
  File "/home/tpiron/maxoled/reseau.py", line 777, in operations
    running = self._first_reaction_method()
  File "/home/tpiron/maxoled/reseau.py", line 679, in _first_reaction_method
    self._remove_move_hole_events(event)
  File "/home/tpiron/maxoled/reseau.py", line 436, in _remove_move_hole_events
    self._new_move_hole_event(item.initial)
  File "/home/tpiron/maxoled/reseau.py", line 473, in _new_move_hole_event
    events = [
  File "/home/tpiron/maxoled/reseau.py", line 474, in <listcomp>
    Event(position, neighbour, self._time_move_hole(position, neighbour), EVENTS["move"], PARTICULES["hole"])
  File "/home/tpiron/maxoled/reseau.py", line 334, in _time_move_hole
    delta_energy += self._hole_electrostatic_energy(initial, final)
  File "/home/tpiron/maxoled/reseau.py", line 355, in _hole_electrostatic_energy
    new_ir : float = self._inverse_radius(((location - final) * self._lattice_constant).norm())
  File "/home/tpiron/maxoled/reseau.py", line 364, in _inverse_radius
    return 1. / distance - 1. / self._cutoff
ZeroDivisionError: float division by zero
```
The `ValueError` raised inside the `_remove_move_hole_events()` method is used to break a while True loop so it is intentional.
However, the `ZeroDivisionError` raised while evaluating the `_inverse_radius()` tells us the substraction `location - final` is the `Vector(0,0,0)`. It is not meant to happen because for `location` to be equal to `final` means that the `_new_move_hole_event()` method allows a forbidden action which is a hole moving to an occupied molecule.

My hypothesis regarding this issue are : 
1. A hole wasn't moved correctly and its position exists inside the `_holes_locations` list allowing the `_hole_electrostatic_energy()` to catch its location. In this case, the fault comes from the `switch_hole()` method of the `Molecule` subclasses and its usage inside the `Lattice` class.
2. Conversely, the position of a moved hole wasn't cleared allowing the `_hole_electrostatic_energy()` to find its location even though the hole doesn't exist. In this case, the fault comes from one of the transformation methods of the `Lattice` class who did not clear the list properly.

In both cases, this problem should be adressed for electrons as well because their behaviors are identical.

_Currently investigating._

**Comment by Pehnny** [comment]
Error happens when an electron tries to move but there is a hole on its location so it should form an exciton. I noticed a bound event in the cache several times indicating this issue may be related to the solved #12 issue. Here is a example of gattered infos : 
```python
Traceback (most recent call last):
  File "/home/tpiron/maxoled/reseau.py", line 792, in operations
    running = self._first_reaction_method()
  File "/home/tpiron/maxoled/reseau.py", line 719, in _first_reaction_method
    self._new_move_event()
  File "/home/tpiron/maxoled/reseau.py", line 496, in _new_move_event
    events : list[Event] = [
  File "/home/tpiron/maxoled/reseau.py", line 497, in <listcomp>
    Event(molecule.position, neighbour, self._time_move_electron(molecule.position, neighbour), EVENTS["move"], PARTICULES["electron"])
  File "/home/tpiron/maxoled/reseau.py", line 296, in _time_move_electron
    delta_energy += self._electron_electrostatic_energy(initial, final)
  File "/home/tpiron/maxoled/reseau.py", line 310, in _electron_electrostatic_energy
    old_ir : float = self._inverse_radius(((location - initial) * self._lattice_constant).norm())
  File "/home/tpiron/maxoled/reseau.py", line 359, in _inverse_radius
    return 1. / distance - 1. / self._cutoff
ZeroDivisionError: float division by zero

deque([
    Event(initial=Point(x=9, y=15, z=15), final=Point(x=9, y=14, z=15), tau=0.0, kind=1, particule=2), 
    Event(initial=Point(x=9, y=14, z=15), final=Point(x=9, y=14, z=15), tau=0.0, kind=2, particule=3), 
    Event(initial=Point(x=9, y=14, z=15), final=Point(x=9, y=15, z=15), tau=0.0, kind=7, particule=1), 
    Event(initial=Point(x=9, y=15, z=15), final=Point(x=9, y=14, z=15), tau=0.0, kind=1, particule=1), 
    Event(initial=Point(x=9, y=14, z=15), final=Point(x=9, y=14, z=15), tau=0.0, kind=2, particule=3), 
    Event(initial=Point(x=3, y=1, z=18), final=Point(x=3, y=0, z=18), tau=0.0, kind=1, particule=1), 
    Event(initial=Point(x=9, y=16, z=16), final=Point(x=9, y=15, z=15), tau=0.0, kind=1, particule=1), 
    Event(initial=Point(x=0, y=2, z=2), final=Point(x=0, y=1, z=2), tau=0.0, kind=1, particule=2), 
    Event(initial=Point(x=9, y=14, z=15), final=Point(x=9, y=15, z=15), tau=0.0, kind=7, particule=1), 
    Event(initial=Point(x=9, y=14, z=15), final=Point(x=9, y=15, z=15), tau=0.0, kind=1, particule=2)
], maxlen=10) 

electrons :  8 [
    Point(x=5, y=10, z=17),
    Point(x=15, y=8, z=13),
    Point(x=18, y=12, z=16), 
    Point(x=6, y=1, z=18), 
    Point(x=2, y=12, z=16), 
    Point(x=3, y=0, z=18), 
    Point(x=9, y=15, z=15), 
    Point(x=9, y=15, z=15)
] 

holes :  7 [
    Point(x=6, y=6, z=2), 
    Point(x=16, y=9, z=3), 
    Point(x=8, y=19, z=3), 
    Point(x=4, y=13, z=1), 
    Point(x=19, y=0, z=7), 
    Point(x=0, y=1, z=2), 
    Point(x=9, y=15, z=15)
] 

excitons :  0 [] 

move events :  0 [] 

exciton events :  0 [] 

capture events :  0 [] 
```
It happened with both kind 1 and 7 (move and unbound) events, always after a kind 2 event. In every observed cases, both an electron and a hole are present in there respective locations lists strongly indicating that a kind 2 (bound) event should happen but doesn't.

I think the issue is due to a removed electron/hole following the displacement of a charge on a molecule occupied by a exciton. This could be a behavior conflict between the `_electrons_location`s attribute and the `molecule.electron` (respectively hole) attribute : 
```python
events : list[Event] = [
    Event(molecule.position, neighbour, self._time_move_electron(molecule.position, neighbour), EVENTS["move"], PARTICULES["electron"])
    for molecule in molecules
    for neighbour in molecule.neighbourhood
    if neighbour not in self._electrons_locations + self._excitons_locations
]

###############

if not molecule.electron and event.final.z != (self._dimension.z - 1) :
    self._new_move_event()
elif molecule.electron :
    self._new_bound_event(event.final)
elif event.final.z == (self._dimension.z - 1) :
    self._new_capture_hole_event(event.final)
```
The `molecule.hole` condition here is stronger than the `not in self._electron_locations` one but we don't use it in the first case because it is slower. However, if an electron or a hole is unintently removed the weakest condition may lead to the `ZeroDivisionError` because the weaker condition sees a charge that is actually not there. 
The reason it happens after a hole move every time may be explained by the order the `_new_move_event()` method creates events, not letting the opportunity to the error to raise in the case of an electron.

In conclusion, this issue should be fixed by issue #12 but further investigations are mandatory.

**Comment by Pehnny** [comment]
The error happened again for a kind 7 event (unbound).

I think it happend when a hole jumps toward another electron. The algorithm tries to create a new move event where a new bound event should.
